### PR TITLE
Use NumPy 1.7 API for pybind11

### PR DESCRIPTION
**Pull request by pschella**

### CODE REVIEW

File: include/ndarray/converter/ufunctors.h [hunk 1]
```diff
@@ -26,17 +26,17 @@ template <typename TUnaryFunctor,
 struct PyUnaryUFunctor {
 
     static PyObject* _call_(TUnaryFunctor const& self, PyObject* input, PyObject* output) {
-        PyObject* input_array = PyArray_FROM_OTF(input,detail::NumpyTraits<TArgument>::getCode(),
-                                                 NPY_ALIGNED);
+        PyArrayObject* input_array = PyArray_FROM_OTF(input,detail::NumpyTraits<TArgument>::getCode(),
+                                                 NPY_ARRAY_ALIGNED);
```
Comment (TallJimbo) [inline include/ndarray/converter/ufunctors.h]: Reindent second line, please.

File: include/ndarray/converter/eigen.h [hunk 2]
```diff
@@ -52,17 +52,17 @@ struct PyConverter< EigenView<T,N,C,XprKind_,Rows_,Cols_> > {
         if (!PyConverter< Array<T,N,C> >::fromPythonStage1(p)) return false;
         // check whether the size is correct if it's static
         if (N == 2) {
-            if (Rows_ != Eigen::Dynamic && PyArray_DIM(p.get(), 0) != Rows_) {
+            if (Rows_ != Eigen::Dynamic && PyArray_DIM(reinterpret_cast<PyArrayObject*>(p.get()), 0) != Rows_) {
```
Comment (TallJimbo) [inline include/ndarray/converter/eigen.h]: Looks like it'd be good to use a temporary `PyArrayObject*` variable here, too.

**Comment by TallJimbo** [comment]
If the user does not define `NPY_NO_DEPRECATED_API` before importing the NumPy headers, do they just get a warning (but everything works?)

If so, maybe the best way to handle this is to put a `#ifndef NPY_NO_DEPRECATED_API` block in one of these headers with *another* warning that tells the user that ndarray is NumPy 1.7-friendly and they can disable both warnings by adding that define.  I don't think we have much choice but to add that `#define` to all of the LSST code using NumPy.

Ultimately, I'd still like to switch all of this code to use the pybind11 NumPy API directly and avoid all the NumPy imports, but that's a much bigger change that will require extensive testing and probably adding some things upstream in pybind11.

**Comment by pschella** [comment]
Yes, that is the case. I added a warning as requested.

**Pull request merged**